### DISCUSSION TITLE
Preventing from scrolling on viewport height

### DISCUSSION THREAD

**Discussion by spacepolice10** [oleggrishechkin/react-viewport-list#52] [Q&A]
Hey,

I'm not sure if there's a way to achieve something like this, but maybe. I have a component with ViewportList which always scrolls to the last element. It receives data twice. First one is cached from previous page. Second one — when additional data loaded. The problem is that additional data prepends the list. So I see how browser quickly gets me to the top of the page and then again at the position I set in ViewportList. 

Maybe there's a way to stop that scroll to top action? Maybe there's a way to create some kind of extreme zero overscan so ViewportList prevents vieweport and DOM from height changes? I want it to stick to last element but without sacrificing loading additional data. Any help really appreciated. Thanks in advance

**Comment by oleggrishechkin** [comment]
Hi, @spacepolice10 
What's a case? As I right understand you describe something like "chat". 

I'll try case with prepending a large data to the list.

Will be good if you provide reproduction in codesandbox for example or some your code snippets

**Reply by oleggrishechkin** [reply]
I created a test demo with prepending a large amount of items to the list and yeah it's flicker slightly on adding + scrolling.
I know why this is happened but I don't know how to avoid it. I store start/end indexes in the list state and when you prepend/unshift items the new items will be showed instead of the old one, because indexes are shifted too.

I thinks it's really hard to implement a hack  on the app side. I think about this case and I'll try to implement a solution for library to avoid this shift.

**Reply by spacepolice10** [reply]
Thank you very much for your help

Ok, for now I will just go with the loader screen before all the data is ready. It is not that bad at the end of the day. Thanks again

**Reply by spacepolice10** [reply]
Also, is there a way to support your project financially?

**Reply by oleggrishechkin** [reply]
> Also, is there a way to support your project financially?

Nope. I don't connected to any donation platform at this moment.
But I think I should get one since somebody wants to support me. Who I am to say no for donat :)

**Reply by oleggrishechkin** [reply]
@spacepolice10 I added buymeacoffee.com and now you can [support me](https://www.buymeacoffee.com/oleggrishechkin) if you wish

**Reply by oleggrishechkin** [reply]
@spacepolice10 as for initial question - for me when I prepend items to the list and then manually call scrollToIndex in useLayoutEffect I see content flushing - some incorrect stuff rendered and then after scrollToIndex logic happened list looks good. Maybe it's now your issue, but another one.

I rewriting scrollToIndex logic in future version to work sync or in useLayoutEffect instead of requestAnimationFrame (current logic). With new logic I don't see any content flushing and see the correct indexes in viewport (maybe because scroll logic run sync with scrollToIndex method call instead of another frame). The new version is not ready now but I want to publish beta soon.

**Reply by spacepolice10** [reply]
> @spacepolice10 I added buymeacoffee.com and now you can [support me](https://www.buymeacoffee.com/oleggrishechkin) if you wish

There is something I would like to discuss regarding to providing you some financial help. Can we contact somewhere else about that? Maybe you happened to live in CIS and there's an option to send some cash directly p2p? If this is not the option I will find a way to use **buymeacoffe**

**Reply by spacepolice10** [reply]
@oleggrishechkin the main problem here is the flashing of prepended content. Also, I believe the same thing preventing from implementation of backward infinite list when you scroll up because if you prepend something browser pushes you to the end of the list. Maybe it is possible to somehow alter index changing behavior, so prepended content won't trigger scrolling in any form?

**Reply by oleggrishechkin** [reply]
> > @spacepolice10 I added buymeacoffee.com and now you can [support me](https://www.buymeacoffee.com/oleggrishechkin) if you wish
> 
> 
> 
> There is something I would like to discuss regarding to providing you some financial help. Can we contact somewhere else about that? Maybe you happened to live in CIS and there's an option to send some cash directly p2p? If this is not the option I will find a way to use **buymeacoffe**

OK, @spacepolice10 you can email to oleggrishechkin@gmail.com

**Reply by spacepolice10** [reply]
> @oleggrishechkin the main problem here is the flashing of prepended content. Also, I believe the same thing preventing from implementation of backward infinite list when you scroll up because if you prepend something browser pushes you to the end of the list. Maybe it is possible to somehow alter index changing behavior, so prepended content won't trigger scrolling in any form?

Have you ever heard about Virtuoso? It is virtualisation library that has an option like `firstItemIndex`. It can be used to change index manually. Would be cool to see something like that in ViewportList, so it solves both problems of prepending items

**Reply by oleggrishechkin** [reply]
> > @oleggrishechkin the main problem here is the flashing of prepended content. Also, I believe the same thing preventing from implementation of backward infinite list when you scroll up because if you prepend something browser pushes you to the end of the list. Maybe it is possible to somehow alter index changing behavior, so prepended content won't trigger scrolling in any form?
> 
> Have you ever heard about Virtuoso? It is virtualisation library that has an option like `firstItemIndex`. It can be used to change index manually. Would be cool to see something like that in ViewportList, so it solves both problems of prepending items 

Yes, I have. I thinking about how to change rendered indexes manually. The problem is you should do scroll compensation at least for safari - because prepended items will shift down items below, but scrollTop will not be changed. So I need make a kind of anchoring to achieve scroll correction.

The best thing - 
Save current rendered item in viewport as anchor, save offset and index. After updating items, find new index of this anchor item - do scrollToIndex. Scroll will be correct for all browsers. No flushing since everything will be called in layoutEffect.

The problem is how to implement API for saving anchor and finding new index in changed items array.
Maybe some callback to return new indexes with default auto-mode for cases like prepending items. firstItemIndex can be a solution too, but looks like it works only for prepending.

I tried in my experiments onItemsChange([oldStartIndex, oldEndIndex]) => [newStartIndex, newEndIndex] callback, and call it during render or in layoutEffect. It partially works.

**Comment by oleggrishechkin** [comment]
I added `indexesShift ` in https://github.com/oleggrishechkin/react-viewport-list/pull/53

It something like firstItemIndex (thank you for reference) but represents difference between old and new start index. You can use any number you want in theory - the main rule is `nextIndexesShift - prevIndexesShift = nextStartIndex - prevStartIndex`

It works for me in Chrome for prepending items but require a little bit client code to update indexesShift for any prepending.

**Comment by oleggrishechkin** [comment]
@spacepolice10 I published beta version of 7.0.0, this release includes indexesShift prop. You can try it. Be careful when upgrading since this 7.0.0 has a several breaking changes

**Reply by spacepolice10** [reply]
Thank you very much. Will try it ASAP. P.S. Did you receive the last email from me? Just in case

**Reply by oleggrishechkin** [reply]
> Thank you very much. Will try it ASAP. P.S. Did you receive the last email from me? Just in case

Yes, everything is ok, thank you for your support. 
(I have some issues with email sending, I don't know why)
